### PR TITLE
[Feat] Add Instrumentation

### PR DESCRIPTION
**Pull request by reyna-abhyankar**
- [x] Adds instrumentation for the following telemetry data
  - Whether the optimizer is run in resume (`-r`) mode
  - Whether the original workflow is written in Langchain, DSPy, or Cognify's programming model
  - Information about the search parameters: 
    - light, medium or heavy search (or an application-specific search)
    - number of trials
    - quality constraint
  - The relative quality/cost improvement over the baseline
- [x] Minor refactoring of workflow translation to allow us to collect workflow info without launching all the optimizer processes

We notify the user on first invocation on the command line. Exporting telemetry data can be turned off by setting `COGNIFY_TELEMETRY=false` as an environment variable.

### CODE REVIEW

File: cognify/_tracing.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+
+def generate_user_identifier():
+    """
+    Generate a consistent, anonymized user identifier.
+    
+    Creates a hash based on:
+    - Hostname
+    - IP address
+    - Current process ID
```
[the remark this answers is not in the document]
Comment (reyna-abhyankar) [reply cognify/_tracing.py]: Removed the process ID (wasn't in the code, but forgot to delete in the comments). It's not needed

File: cognify/_tracing.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+is_telemetry_on = os.getenv("COGNIFY_TELEMETRY", "true").lower() == "true"
+
+resource = Resource.create(attributes={
+    "service.name": "cognify",
+    "user.id": generate_user_identifier()
+})
+
+provider = TracerProvider(resource=resource)
+# processor = BatchSpanProcessor(ConsoleSpanExporter())
+processor = BatchSpanProcessor(OTLPSpanExporter(endpoint="http://wuklab-01.ucsd.edu:4318/v1/traces"))
```
[the remark this answers is not in the document]
Comment (reyna-abhyankar) [reply cognify/_tracing.py:78]: Added a comment saying they can add an exporter for custom telemetry handling

File: cognify/_tracing.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    # Define the marker file path
+    marker_file_name = f".cognify_{version.replace('.', '_')}_initial_run"
+    marker_file = Path(__file__).parent / marker_file_name
+
+    if not marker_file.exists():
+        # Show the first-run message
+        print(f"""
+Thank you for using Cognify-{version}! 🚀 To better understand how people use Cognify, we collect the following telemetry data:
+    - Whether the optimizer is run in `resume` (-r) mode
+    - Whether the original workflow is written in Langchain, DSPy, or Cognify's programming model
```
[the remark this answers is not in the document]
Comment (reyna-abhyankar) [reply cognify/_tracing.py:28]: Done

File: cognify/_tracing.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+def initial_usage_message():
+    version = pkg_resources.get_distribution('cognify-ai').version
+    # Define the marker file path
+    marker_file_name = f".cognify_{version.replace('.', '_')}_initial_run"
+    marker_file = Path(__file__).parent / marker_file_name
+
+    if not marker_file.exists():
+        # Show the first-run message
+        print(f"""
+Thank you for using Cognify-{version}! 🚀 To better understand how people use Cognify, we collect the following telemetry data:
```
[the remark this answers is not in the document]
Comment (reyna-abhyankar) [reply cognify/_tracing.py]: Done

File: cognify/_tracing.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yiying-zhang) [reply cognify/_tracing.py:79]: where's this?

[the remark this answers is not in the document]
Comment (reyna-abhyankar) [reply cognify/_tracing.py:79]: the next line here (line 80): 
`# to set up your own telemetry, you can add a new exporter to a custom endpoint`

**Pull request merged**